// @devvit/payments/helpers/ProductButton.js
import { Devvit } from '@devvit/public-api';
/**
 * A button that allows users to purchase a product.
 * @param {ProductButtonProps} props - The props for the ProductButton component.
 * @returns {JSX.Element} The ProductButton component.
 */
export function ProductButton({ onPress, product, appearance = 'compact', showIcon = false, buttonAppearance = 'primary', textColor, ...rest }) {
    const isCompact = appearance === 'compact';
    const isTile = appearance === 'tile';
    const icon = product.images?.icon;
    const shouldShowIcon = showIcon && icon;
    const content = (Devvit.createElement(Devvit.Fragment, null,
        shouldShowIcon ? (Devvit.createElement(Devvit.Fragment, null,
            Devvit.createElement(ProductIcon, { url: icon, size: isCompact ? 'small' : 'large' }),
            Devvit.createElement("spacer", { size: "xsmall" }))) : null,
        // Nudge the text away from the edge if there's no icon and it's not a tile.
        !isTile ? Devvit.createElement("spacer", { size: "xsmall" }) : null,
        Devvit.createElement("vstack", { grow: true },
            Devvit.createElement("text", { weight: "bold", color: textColor, grow: false }, product.displayName),
            !isCompact ? (Devvit.createElement("text", { size: "xsmall", color: textColor, wrap: true }, product.description)) : null),
        Devvit.createElement("spacer", { size: isTile ? 'small' : 'medium' }),
        Devvit.createElement("button", { onPress: () => onPress(product), appearance: buttonAppearance, icon: "gold-fill", size: "small" },
            "Use ",
            product.price)));
    const parentStackProps = {
        cornerRadius: isCompact ? 'full' : 'medium',
        gap: 'none',
        border: 'thin',
        padding: 'small',
        alignment: 'start middle',
        ...rest,
    };
    if (appearance === 'tile') {
        return Devvit.createElement("vstack", { ...parentStackProps }, content);
    }
    return Devvit.createElement("hstack", { ...parentStackProps }, content);
}
function ProductIcon({ url, size }) {
    const iconSize = size === 'small' ? '32px' : '50px';
    return (Devvit.createElement("vstack", { cornerRadius: size === 'small' ? 'full' : 'small', height: iconSize, width: iconSize },
        Devvit.createElement("image", { url: url, imageWidth: iconSize, imageHeight: iconSize, resizeMode: "fill" })));
}
