// @devvit/payments/server/PaymentsClient.js
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _PaymentsClient_instances, _PaymentsClient_plugin_get;
import { OrderStatus } from '@devvit/protos/json/devvit/payments/v1alpha/order.js';
import { PaymentsServiceDefinition, } from '@devvit/protos/types/devvit/plugin/payments/v1alpha/payments.js';
import { context } from '@devvit/server';
import { getDevvitConfig } from '@devvit/shared-types/server/get-devvit-config.js';
/**
 * @experimental - This is not finalized yet, may not work, and may change.
 */
export class PaymentsClient {
    constructor() {
        _PaymentsClient_instances.add(this);
    }
    /**
     * @param filters - Filters to apply to the products query.
     * @experimental - This is not finalized yet, may not work, and may change.
     */
    getProducts(filters = {}) {
        return __classPrivateFieldGet(this, _PaymentsClient_instances, "a", _PaymentsClient_plugin_get).GetProducts({ metadata: filters.metadata ?? {}, skus: filters.skus ?? [] }, context.metadata);
    }
    /**
     * @param filters - Filters to apply to the orders query.
     * @experimental - This is not finalized yet, may not work, and may change.
     */
    async getOrders(filters) {
        const orders = await __classPrivateFieldGet(this, _PaymentsClient_instances, "a", _PaymentsClient_plugin_get).GetOrders({
            metadata: filters.metadata ?? {},
            start: filters.start ? new Date(filters.start) : undefined,
            end: filters.end ? new Date(filters.end) : undefined,
            sku: filters.sku ?? '',
            buyer: filters.buyer ?? '',
            status: filters.status ?? OrderStatus.ORDER_STATUS_UNSPECIFIED,
            cursor: filters.cursor ?? '',
            limit: filters.limit,
        }, context.metadata);
        // Convert the hydrated old Protobuf to JSON. Don't use
        // Protobuf.toJSON() which would omit default values.
        return JSON.parse(JSON.stringify(orders));
    }
    /**
     * @param orderId - The ID of the order to acknowledge delivery for.
     * @experimental - This is not finalized yet, may not work, and may change.
     */
    acknowledgeOrderDelivery(orderId) {
        return __classPrivateFieldGet(this, _PaymentsClient_instances, "a", _PaymentsClient_plugin_get).AcknowledgeOrderDelivery({ orderId }, context.metadata);
    }
}
_PaymentsClient_instances = new WeakSet(), _PaymentsClient_plugin_get = function _PaymentsClient_plugin_get() {
    return getDevvitConfig().use(PaymentsServiceDefinition);
};
