// @devvit/payments/hooks/use-orders.js
import { useAsync, useState } from '@devvit/public-api';
import { orderFromProto } from '@devvit/shared-types/payments/Order.js';
import { paymentsPlugin } from '../plugin.js';
export async function getOrders(opts = {}) {
    const request = {
        sku: '',
        buyer: '',
        metadata: {},
        status: 0,
        ...opts,
        limit: opts.limit || 100,
        cursor: opts.nextPageToken || '',
    };
    const ordersResponse = await paymentsPlugin.GetOrders(request);
    const ordersOut = {
        orders: ordersResponse.orders.map(orderFromProto),
    };
    if (ordersResponse.pageInfo?.hasNextPage) {
        ordersOut.nextPage = ordersResponse.pageInfo?.endCursor;
    }
    return ordersOut;
}
function orderToJSON(order) {
    return {
        ...order,
        createdAt: order.createdAt?.toISOString() ?? null,
        updatedAt: order.updatedAt?.toISOString() ?? null,
    };
}
function orderFromJSON(order) {
    return {
        ...order,
        createdAt: order.createdAt ? new Date(order.createdAt) : null,
        updatedAt: order.updatedAt ? new Date(order.updatedAt) : null,
    };
}
export function useOrders(_ctx, opts = {}) {
    const [pageToken, setPageToken] = useState(opts.nextPageToken ?? '');
    const { loading, data, error } = useAsync(async () => {
        const fetchRes = await getOrders({ ...opts, nextPageToken: pageToken });
        return {
            orders: fetchRes.orders.map(orderToJSON),
            nextPageToken: fetchRes.nextPage || '',
        };
    }, {
        depends: pageToken,
    });
    let base = {
        orders: (data?.orders ?? []).map(orderFromJSON),
    };
    if (data?.nextPageToken) {
        base = Object.assign(base, {
            nextPage: async () => {
                setPageToken(data?.nextPageToken);
            },
        });
    }
    if (loading) {
        return {
            ...base,
            loading: true,
            error: null,
        };
    }
    if (error) {
        return {
            ...base,
            loading,
            error,
        };
    }
    return {
        ...base,
        loading,
        error: null,
    };
}
