// @devvit/payments/hooks/use-products.js
import { useAsync, useState } from '@devvit/public-api';
import { productFromProto } from '@devvit/shared-types/payments/Product.js';
import { paymentsPlugin } from '../plugin.js';
export async function getProducts(opts = {}) {
    const request = {
        skus: opts.skus || [],
        metadata: opts.metadata || {},
    };
    return (await paymentsPlugin.GetProducts(request)).products.map(productFromProto);
}
export function useProducts(_ctx, opts = {}) {
    // internal state used to trigger refetch
    const [refetchBit, setRefetchBit] = useState(0);
    const { loading, data, error } = useAsync(async () => getProducts(opts), {
        depends: refetchBit,
    });
    const products = data ?? [];
    const refetch = async () => {
        setRefetchBit((b) => ~b); // flip bit to trigger callback execution
    };
    const base = {
        products,
        refetch,
    };
    if (loading) {
        return {
            ...base,
            loading: true,
            error: null,
        };
    }
    if (error != null) {
        return {
            ...base,
            loading: false,
            error,
        };
    }
    return {
        ...base,
        loading: false,
        error: null,
    };
}
