// @devvit/media/test/mocks/MediaMock.js
export class MediaPluginMock {
    constructor(store) {
        this._store = store;
    }
    async Upload(request, _metadata) {
        this._store.uploads.push(request);
        const mediaId = `media-${this._store.uploads.length}`;
        const extensionMap = {
            image: 'png',
            video: 'mp4',
            gif: 'gif',
        };
        const ext = extensionMap[request.type] ?? 'png';
        // Fake URL generation
        const mediaUrl = `https://i.redd.it/bogus-for-testing/${mediaId}.${ext}`;
        return {
            mediaId,
            mediaUrl,
        };
    }
}
export class MediaMock {
    constructor() {
        this._store = {
            uploads: [],
        };
        this.plugin = new MediaPluginMock(this._store);
    }
    /**
     * Returns the list of media upload requests that have been made.
     */
    get uploads() {
        return this._store.uploads;
    }
    /**
     * Clears all recorded uploads.
     */
    clear() {
        this._store.uploads = [];
    }
}
