// @devvit/reddit/mocks/GraphQLMock.js
const methodMap = {
    GetUserSocialLinks: 'User.getSocialLinks',
    GetSnoovatarUrlByName: 'User.getSnoovatarUrl',
    GetVaultContactByAddress: 'getVaultByAddress',
    GetVaultContactByUserId: 'getVaultByUserId',
    CreateModmailConversation: 'ModMailService.create*Conversation',
    GetSubredditInfoByName: 'getSubredditInfoByName',
    GetSubredditInfoById: 'getSubredditInfoById',
    CreateShareUrl: 'createShareUrl',
};
export class GraphQLPluginMock {
    constructor(linksAndCommentsMock) {
        this._linksAndCommentsMock = linksAndCommentsMock;
    }
    async PersistedQuery(request, _metadata) {
        const { operationName, variables } = request;
        if (operationName === 'GetDevvitPostData') {
            const id = variables?.id;
            if (!id) {
                throw new Error('GetDevvitPostData requires an id');
            }
            const postData = this._linksAndCommentsMock.getPostData(id);
            return {
                data: {
                    postInfoById: {
                        devvit: {
                            postData: postData ? JSON.stringify({ developerData: postData }) : undefined,
                        },
                    },
                },
                errors: [],
            };
        }
        const friendlyName = methodMap[operationName];
        const nameToReport = friendlyName ?? 'unknown';
        throw new Error(`Reddit API method ${nameToReport} is not implemented in the test harness.\n` +
            `For more information, visit https://developers.reddit.com/docs/guides/tools/devvit_test`);
    }
    async Query(_request, _metadata) {
        throw new Error(`Reddit API method GraphQL.Query is not implemented in the test harness.\n` +
            `For more information, visit https://developers.reddit.com/docs/guides/tools/devvit_test`);
    }
}
export class GraphQLMock {
    constructor(linksAndCommentsMock) {
        this.plugin = new GraphQLPluginMock(linksAndCommentsMock);
    }
}
