// @devvit/reddit/mocks/SubredditMock.js
export class SubredditPluginMock {
    constructor(store) {
        this._store = store;
    }
    async AboutWhere(_request, _metadata) {
        throw new Error(`Reddit API method Subreddits.AboutWhere is not implemented in the test harness.\n` +
            `For more information, visit https://developers.reddit.com/docs/guides/tools/devvit_test`);
    }
    async DeleteSrBanner(_request, _metadata) {
        throw new Error(`Reddit API method Subreddits.DeleteSrBanner is not implemented in the test harness.\n` +
            `For more information, visit https://developers.reddit.com/docs/guides/tools/devvit_test`);
    }
    async DeleteSrHeader(_request, _metadata) {
        throw new Error(`Reddit API method Subreddits.DeleteSrHeader is not implemented in the test harness.\n` +
            `For more information, visit https://developers.reddit.com/docs/guides/tools/devvit_test`);
    }
    async DeleteSrIcon(_request, _metadata) {
        throw new Error(`Reddit API method Subreddits.DeleteSrIcon is not implemented in the test harness.\n` +
            `For more information, visit https://developers.reddit.com/docs/guides/tools/devvit_test`);
    }
    async DeleteSrImg(_request, _metadata) {
        throw new Error(`Reddit API method Subreddits.DeleteSrImg is not implemented in the test harness.\n` +
            `For more information, visit https://developers.reddit.com/docs/guides/tools/devvit_test`);
    }
    async SearchRedditNames(_request, _metadata) {
        throw new Error(`Reddit API method Subreddits.SearchRedditNames is not implemented in the test harness.\n` +
            `For more information, visit https://developers.reddit.com/docs/guides/tools/devvit_test`);
    }
    async SearchSubreddits(_request, _metadata) {
        throw new Error(`Reddit API method Subreddits.SearchSubreddits is not implemented in the test harness.\n` +
            `For more information, visit https://developers.reddit.com/docs/guides/tools/devvit_test`);
    }
    async Sidebar(_request, _metadata) {
        throw new Error(`Reddit API method Subreddits.Sidebar is not implemented in the test harness.\n` +
            `For more information, visit https://developers.reddit.com/docs/guides/tools/devvit_test`);
    }
    async SiteAdmin(_request, _metadata) {
        throw new Error(`Reddit API method Subreddits.SiteAdmin is not implemented in the test harness.\n` +
            `For more information, visit https://developers.reddit.com/docs/guides/tools/devvit_test`);
    }
    async Sticky(_request, _metadata) {
        throw new Error(`Reddit API method Subreddits.Sticky is not implemented in the test harness.\n` +
            `For more information, visit https://developers.reddit.com/docs/guides/tools/devvit_test`);
    }
    async SubmitText(_request, _metadata) {
        throw new Error(`Reddit API method Subreddits.SubmitText is not implemented in the test harness.\n` +
            `For more information, visit https://developers.reddit.com/docs/guides/tools/devvit_test`);
    }
    async SubredditAbout(request, _metadata) {
        const subredditName = request.subreddit;
        let found;
        for (const [name, data] of this._store.subreddits.entries()) {
            if (name.toLowerCase() === subredditName.toLowerCase()) {
                found = data;
                break;
            }
        }
        if (!found) {
            throw new Error('HTTP 404 Not Found');
        }
        return { data: found };
    }
    async SubredditAboutEdit(_request, _metadata) {
        throw new Error(`Reddit API method Subreddits.SubredditAboutEdit is not implemented in the test harness.\n` +
            `For more information, visit https://developers.reddit.com/docs/guides/tools/devvit_test`);
    }
    async SubredditAboutRules(_request, _metadata) {
        throw new Error(`Reddit API method Subreddits.SubredditAboutRules is not implemented in the test harness.\n` +
            `For more information, visit https://developers.reddit.com/docs/guides/tools/devvit_test`);
    }
    async SubredditAboutTraffic(_request, _metadata) {
        throw new Error(`Reddit API method Subreddits.SubredditAboutTraffic is not implemented in the test harness.\n` +
            `For more information, visit https://developers.reddit.com/docs/guides/tools/devvit_test`);
    }
    async SubredditAddRemovalReason(_request, _metadata) {
        throw new Error(`Reddit API method Subreddits.SubredditAddRemovalReason is not implemented in the test harness.\n` +
            `For more information, visit https://developers.reddit.com/docs/guides/tools/devvit_test`);
    }
    async SubredditAutocomplete(_request, _metadata) {
        throw new Error(`Reddit API method Subreddits.SubredditAutocomplete is not implemented in the test harness.\n` +
            `For more information, visit https://developers.reddit.com/docs/guides/tools/devvit_test`);
    }
    async SubredditGetRemovalReasons(_request, _metadata) {
        throw new Error(`Reddit API method Subreddits.SubredditGetRemovalReasons is not implemented in the test harness.\n` +
            `For more information, visit https://developers.reddit.com/docs/guides/tools/devvit_test`);
    }
    async SubredditPostRequirements(_request, _metadata) {
        throw new Error(`Reddit API method Subreddits.SubredditPostRequirements is not implemented in the test harness.\n` +
            `For more information, visit https://developers.reddit.com/docs/guides/tools/devvit_test`);
    }
    async SubredditStylesheet(_request, _metadata) {
        throw new Error(`Reddit API method Subreddits.SubredditStylesheet is not implemented in the test harness.\n` +
            `For more information, visit https://developers.reddit.com/docs/guides/tools/devvit_test`);
    }
    async SubredditsMineWhere(_request, _metadata) {
        throw new Error(`Reddit API method Subreddits.SubredditsMineWhere is not implemented in the test harness.\n` +
            `For more information, visit https://developers.reddit.com/docs/guides/tools/devvit_test`);
    }
    async SubredditsSearch(_request, _metadata) {
        throw new Error(`Reddit API method Subreddits.SubredditsSearch is not implemented in the test harness.\n` +
            `For more information, visit https://developers.reddit.com/docs/guides/tools/devvit_test`);
    }
    async SubredditsWhere(_request, _metadata) {
        throw new Error(`Reddit API method Subreddits.SubredditsWhere is not implemented in the test harness.\n` +
            `For more information, visit https://developers.reddit.com/docs/guides/tools/devvit_test`);
    }
    async Subscribe(_request, _metadata) {
        throw new Error(`Reddit API method Subreddits.Subscribe is not implemented in the test harness.\n` +
            `For more information, visit https://developers.reddit.com/docs/guides/tools/devvit_test`);
    }
    async UploadSrImg(_request, _metadata) {
        throw new Error(`Reddit API method Subreddits.UploadSrImg is not implemented in the test harness.\n` +
            `For more information, visit https://developers.reddit.com/docs/guides/tools/devvit_test`);
    }
    async UsersSearch(_request, _metadata) {
        throw new Error(`Reddit API method Subreddits.UsersSearch is not implemented in the test harness.\n` +
            `For more information, visit https://developers.reddit.com/docs/guides/tools/devvit_test`);
    }
    async UsersWhere(_request, _metadata) {
        throw new Error(`Reddit API method Subreddits.UsersWhere is not implemented in the test harness.\n` +
            `For more information, visit https://developers.reddit.com/docs/guides/tools/devvit_test`);
    }
}
export class SubredditMock {
    constructor() {
        this._store = {
            subreddits: new Map(),
        };
        this.plugin = new SubredditPluginMock(this._store);
    }
    /**
     * Helper to seed the mock database with a Subreddit.
     * This allows tests to set up state before calling `reddit.getSubredditByName`.
     */
    addSubreddit(data) {
        const subreddit = {
            createdUtc: Math.floor(Date.now() / 1000),
            subscribers: 0,
            activeUserCount: 0,
            over18: false,
            lang: 'en',
            url: `/r/${data.name}/`,
            title: data.displayName,
            description: '',
            subredditType: 'public',
            ...data,
            id: data.id.replace(/^t5_/, ''),
        };
        this._store.subreddits.set(data.displayName, subreddit);
        return subreddit;
    }
}
