// @devvit/settings/mocks/SettingsMock.js
import {} from '@devvit/protos';
import { toFormFieldValue } from '../SettingsClient.js';
export class SettingsPluginMock {
    constructor(state) {
        this._state = state;
    }
    async GetSettings(_request, _metadata) {
        return {
            appSettings: this._state.current,
            installationSettings: { version: '0', settings: {} },
            instanceSettings: { version: '0', settings: {} },
        };
    }
}
export class SettingsMock {
    constructor(initialSettings = {}) {
        this._state = { current: this._createSettingsValues(initialSettings) };
        this.plugin = new SettingsPluginMock(this._state);
    }
    /**
     * Overwrites the mocked settings with new values.
     * Replaces existing settings completely.
     *
     * @param settings - The new settings to replace the current ones.
     */
    set(settings) {
        const newValues = this._createSettingsValues(settings);
        this._state.current = newValues;
    }
    /**
     * Updates the mocked settings with new values.
     * Merges with existing settings.
     *
     * @param settings - The new settings to merge in.
     */
    update(settings) {
        const newValues = this._createSettingsValues(settings);
        this._state.current = {
            ...this._state.current,
            settings: { ...this._state.current.settings, ...newValues.settings },
        };
    }
    /**
     * Adds or updates a single setting value.
     *
     * @param key - The setting key.
     * @param value - The value to set.
     */
    put(key, value) {
        this.update({ [key]: value });
    }
    /**
     * Removes a setting by key.
     *
     * @param key - The setting key to remove.
     */
    remove(key) {
        delete this._state.current.settings[key];
    }
    /**
     * Gets the current settings values.
     *
     * @returns The current SettingsValues object.
     */
    get() {
        return this._state.current;
    }
    _createSettingsValues(settings) {
        const formFields = {};
        for (const [key, value] of Object.entries(settings)) {
            formFields[key] = toFormFieldValue(value);
        }
        return {
            version: '1',
            settings: formFields,
        };
    }
}
