// @devvit/redis/redisCompression.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _RedisCompressionProxy_client;
import { gunzipSync, gzipSync } from 'node:zlib';
import { RedisKeyScope, } from '@devvit/protos/json/devvit/plugin/redis/redisapi.js';
import { RedisClient as RedisClientImpl } from './RedisClient.js';
// Envelope parameters
const COMPRESSION_ALGO = 'gz';
const ENCODING_ALGO = 'b64';
// Envelope format: __<compression>:<encoding>__:
// We use a structured envelope to allow for future changes in compression/encoding
// while keeping the prefix small and unlikely to collide with user data.
export const REDIS_COMPRESSION_PREFIX = `__${COMPRESSION_ALGO}:${ENCODING_ALGO}__:`;
// Set a safe threshold where compression is unlikely to be beneficial.
const MIN_COMPRESSION_LENGTH = 80;
const compress = (value) => {
    if (value.length < MIN_COMPRESSION_LENGTH)
        return value;
    try {
        const compressed = gzipSync(value);
        return `${REDIS_COMPRESSION_PREFIX}${compressed.toString('base64')}`;
    }
    catch {
        return value;
    }
};
const decompress = (value) => {
    if (!value.startsWith(REDIS_COMPRESSION_PREFIX))
        return value;
    try {
        const buffer = Buffer.from(value.slice(REDIS_COMPRESSION_PREFIX.length), 'base64');
        return gunzipSync(buffer).toString('utf-8');
    }
    catch {
        return value;
    }
};
export class RedisCompressionProxy {
    constructor(client) {
        _RedisCompressionProxy_client.set(this, void 0);
        __classPrivateFieldSet(this, _RedisCompressionProxy_client, client, "f");
        this.global =
            client.scope === RedisKeyScope.INSTALLATION
                ? new RedisCompressionProxy(client.global)
                : this;
    }
    async get(key) {
        const val = await __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").get(key);
        if (!val)
            return val;
        return val.startsWith(REDIS_COMPRESSION_PREFIX) ? decompress(val) : val;
    }
    async set(key, value, options) {
        const compressed = compress(value);
        const toStore = compressed.length < value.length ? compressed : value;
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").set(key, toStore, options);
    }
    async hGet(key, field) {
        const val = await __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").hGet(key, field);
        if (!val)
            return val;
        return val.startsWith(REDIS_COMPRESSION_PREFIX) ? decompress(val) : val;
    }
    async hSet(key, fieldValues) {
        const newFieldValues = {};
        for (const [field, value] of Object.entries(fieldValues)) {
            const compressed = compress(value);
            newFieldValues[field] = compressed.length < value.length ? compressed : value;
        }
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").hSet(key, newFieldValues);
    }
    async hSetNX(key, field, value) {
        const compressed = compress(value);
        const toStore = compressed.length < value.length ? compressed : value;
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").hSetNX(key, field, toStore);
    }
    async hGetAll(key) {
        const all = await __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").hGetAll(key);
        if (!all) {
            return all;
        }
        const result = {};
        for (const [field, value] of Object.entries(all)) {
            result[field] = value.startsWith(REDIS_COMPRESSION_PREFIX) ? decompress(value) : value;
        }
        return result;
    }
    async hMGet(key, fields) {
        const values = await __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").hMGet(key, fields);
        return values.map((val) => {
            if (val?.startsWith(REDIS_COMPRESSION_PREFIX)) {
                return decompress(val);
            }
            return val;
        });
    }
    async mGet(keys) {
        const values = await __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").mGet(keys);
        return values.map((val) => {
            if (val?.startsWith(REDIS_COMPRESSION_PREFIX)) {
                return decompress(val);
            }
            return val;
        });
    }
    async mSet(keyValues) {
        const newKeyValues = {};
        for (const [key, value] of Object.entries(keyValues)) {
            const compressed = compress(value);
            newKeyValues[key] = compressed.length < value.length ? compressed : value;
        }
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").mSet(newKeyValues);
    }
    watch(...keys) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").watch(...keys);
    }
    getBuffer(key) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").getBuffer(key);
    }
    exists(...keys) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").exists(...keys);
    }
    del(...keys) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").del(...keys);
    }
    type(key) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").type(key);
    }
    rename(key, newKey) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").rename(key, newKey);
    }
    getRange(key, start, end) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").getRange(key, start, end);
    }
    setRange(key, offset, value) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").setRange(key, offset, value);
    }
    strLen(key) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").strLen(key);
    }
    incrBy(key, value) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").incrBy(key, value);
    }
    expire(key, seconds) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").expire(key, seconds);
    }
    expireTime(key) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").expireTime(key);
    }
    zAdd(key, ...members) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").zAdd(key, ...members);
    }
    zCard(key) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").zCard(key);
    }
    zScore(key, member) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").zScore(key, member);
    }
    zRank(key, member) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").zRank(key, member);
    }
    zIncrBy(key, member, value) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").zIncrBy(key, member, value);
    }
    zRange(key, start, stop, options) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").zRange(key, start, stop, options);
    }
    zRem(key, members) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").zRem(key, members);
    }
    zRemRangeByLex(key, min, max) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").zRemRangeByLex(key, min, max);
    }
    zRemRangeByRank(key, start, stop) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").zRemRangeByRank(key, start, stop);
    }
    zRemRangeByScore(key, min, max) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").zRemRangeByScore(key, min, max);
    }
    zScan(key, cursor, pattern, count) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").zScan(key, cursor, pattern, count);
    }
    hDel(key, fields) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").hDel(key, fields);
    }
    hScan(key, cursor, pattern, count) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").hScan(key, cursor, pattern, count);
    }
    hKeys(key) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").hKeys(key);
    }
    hIncrBy(key, field, value) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").hIncrBy(key, field, value);
    }
    hLen(key) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").hLen(key);
    }
    bitfield(key, ...cmds) {
        return __classPrivateFieldGet(this, _RedisCompressionProxy_client, "f").bitfield(key, ...cmds);
    }
}
_RedisCompressionProxy_client = new WeakMap();
